### ISSUE TITLE
[RFC] Add JIT graph fuser for oneDNN Graph API

### ISSUE THREAD

**Issue by pinzhenx** [pytorch/pytorch#49444]
## Feature

This RFC proposes to add the infrastructure of a new PyTorch JIT graph fuser based on [oneDNN Graph API](https://spec.oneapi.com/onednn-graph/latest/introduction.html) (codename: LLGA).

**oneDNN Graph API** _extends_ oneDNN with a high-level graph API, as a complement to the **oneDNN Primitive API**. It can benefit the PyTorch community by maximizing the optimization opportunity with target-specific optimization and code generation in a larger scope than individual DL ops across a variety of Intel and non-Intel HW.

oneDNN Graph API accepts a deep learning computation graph based on its [Opset](https://spec.oneapi.com/onednn-graph/latest/ops/index.html) as input and performs graph partitioning, where nodes that are candidates for fusion are grouped together. oneDNN Graph compiles and executes a group of deep learning operations in a graph partition as a fused operation. [oneDNN Graph programming model](https://spec.oneapi.com/onednn-graph/latest/programming_model.html) is designed for easy integration with deep learning frameworks.

The PyTorch integration will be mainly composed of two parts: 1) **Graph rewrite** for graph partitioning and fusion; 2) **Runtime** for partition compilation and execution.

1. Graph Rewrite

    oneDNN Graph rewrite pass applies fusion on TorchScript IR according to oneDNN Graph partitioning. The rewrite pass is composed of the following stages:

    - **Passing graph**: Translate TorchScript IR into oneDNN Graph IR represented by the `graph` object. All PyTorch `Node`s are mapped to oneDNN Graph `op`s and passed with `graph::add_op(op)`.

    - **Partitioning**: oneDNN Graph picks out regions of interest and returns `std::vector<partition>`. Each `partition` corresponds to an unordered list of ops to form a fusion op.

    - **Rewrite**: Cluster JIT nodes into fusion ops based on the partitions returned from the partitioning stage.

2. Runtime

    Each oneDNN Graph partition is [compiled](https://spec.oneapi.com/onednn-graph/latest/programming_model.html#partition) and [executed](https://spec.oneapi.com/onednn-graph/latest/programming_model.html#compiled-partition) in the implementation of the fusion op.

## Pitch

Add the infrastructure of a new JIT graph fuser on top of oneDNN Graph API.

## Additional context

We are structuring a new graph fuser, that follows many of the designs of TensorExpression and CudaFuser.

1. PyTorch JIT Integration and User Opt-in API

    Most of the source code lives in `torch/csrc/jit/codegen/onednn/`. Since oneDNN Graph supports fusion, code generation and execution with a unified graph API, it results in a very light-weight bridge code (< 2k LOC) for the new fuser.

    We plan to register a new custom pre-pass for graph rewrite and a new JIT op called `prim::LlgaFusionGroup` for runtime implementation. The optimization pass will be disabled by default at first, and users should enable it by calling

    ```python
    torch._C._jit_set_llga_enabled(True)
    ```

2. Alias and Mutation Reduction

    oneDNN Graph Ops are designed to be pure functional in order to simplify the graph analysis. PyTorch Ops with alias and mutation semantics are treated as [unsupported Ops](https://spec.oneapi.com/onednn-graph/latest/ops/misc/Wildcard_1.html) and will be excluded from the fusion. In order to maximize the fusion opportunities, there has to be a graph rewrite pass beforehand to convert PyTorch Ops with alias and mutation semantics to Ops with functional semantics with the best effort. We notice that there is a `MutationRemover` utilily class that can help. We are also aware that PyTorch is introducing the functional transformation (FX) tool to facilitate such a graph rewrite pass. We want to ask for architectural guidance on what is the recommended path to follow.

    oneDNN Graph API does not support control flow yet and the current integration only supports tracing graphs that do not have control flow ops.

@jgong5 @Jianhui-li

**Comment by Jianhui-Li** [comment]
@gottbrath

**Comment by pinzhenx** [comment]
Preview branch of the PR: https://github.com/pytorch/pytorch/compare/v1.7.1-rc2...pinzhenx:llga-preview

Doc [here](https://github.com/pinzhenx/pytorch/blob/llga-preview/torch/csrc/jit/codegen/onednn/README.md)

**Comment by eellison** [comment]
Cool! 

I have some initial questions: 

What information does the OneDNN fuser specialize on statically during compilation,  and what information does the OneDNN fuser specialize on at invocation? For example, the NVidia fuser during compile time specializes on the # of dimensions and dtype, and on invocation further specializes based on the contiguity of inputs. Is there any limit to the number of specialized kernels that OneDNN will generate ? 

Do you have any examples of models where you were observed speed up using OneDNN? 

Do you have any examples of models where you observed slowdown using OneDNN? 

To what extent does  OneDNN cover all the wide range of tensors you might see in pytorch ? What dtypes does it not handle, what devices, sparsity, strides ? It doesn't need to handle every case but it would be good to know the coverage. 

I think if we can demonstrate perf wins without corresponding degradations, we should merge this, particular for instances of fusion which TensorExprKernel  does not already cover (currrently, any elementwise fusions).

**Comment by asuhan** [comment]
I second the question about benchmark suite, it's important for us to know what to expect and on which models.

**Comment by eellison** [comment]
The benchmark suite may not fully what the speedup could could  be. For instance, in scripted models, conv->batchnorm won't fuse right now because of an if statement, but that could be worked around.

**Comment by pinzhenx** [comment]
@eellison Thanks for your reply.

> What information does the OneDNN fuser specialize on statically during compilation, and what information does the OneDNN fuser specialize on at invocation? For example, the NVidia fuser during compile time specializes on the # of dimensions and dtype, and on invocation further specializes based on the contiguity of inputs.

Same as CudaFuser. Since our design is based on the legacy executor (code is compatible with the profiling executor), we need the dimensionality and dtype only during the static graph rewrite pass. The kernel will then be compiled in the first invocation and be fulfilled with rich tensor details, such as sizes and strides.

<br>

> Is there any limit to the number of specialized kernels that OneDNN will generate?

No such limit from the library. Kernel objects and graph objects are bound together. We are not using a cache to store kernels as it's hard to maintain the lifetime of the kernel objects. So the #kernels should depend on the scale of the graph.

<br>
 
> Do you have any examples of models where you were observed speed up using OneDNN?
> Do you have any examples of models where you observed slowdown using OneDNN?
> Could you run the pytorch benchmark suite and report the results? https://github.com/pytorch/benchmark?

Thanks for providing this suite. We will post the detailed results in our official PR.

We are now targeting CNN models like resne(x)t, and the MLP part in non-CNN models like DLRM. In CNN, we have fusions like `conv-sum`, that's what pytorch don't have right now, and it seems that FX team is attempting to enable it.

<br>

> To what extent does OneDNN cover all the wide range of tensors you might see in pytorch ? What dtypes does it not handle, what devices, sparsity, strides? It doesn't need to handle every case but it would be good to know the coverage.

For strides, what the bridge code does is merely to pass the sizes and strides to the oneDNN graph library. If oneDNN Graph is more favorable of dense tensors for some ops, it will convert the non-contiguous boundary tensors inside the library.

For the device, we are working on CPU only, and we are planning to interoperate with the `XPU` (name is TBD, see [RFC](https://github.com/pytorch/pytorch/issues/48246) here) device in the future.

For dtype, we are now supporting FP32/BF16 tensors.

For sparse tensor, we don't support it yet.

**Comment by pinzhenx** [comment]
@eellison 
I've slightly changed the benchmark code to trace the model (freezing then scripting does not work for us). Here are some initial results collected from the benchmark suite

```bash
# SkyLake 8180, single thread
| name                                             | base time (s) |   diff time (s) |   % change |
| :---                                             |          ---: |            ---: |       ---: |
| test_eval[alexnet-cpu-jit]                       |      0.030493 |        0.025186 |     -17.4% |
| test_eval[densenet121-cpu-jit]                   |      0.146203 |        0.084092 |     -42.5% |
| test_eval[resnet50-cpu-jit]                      |      0.121597 |        0.052696 |     -56.7% |
| test_eval[resnext50_32x4d-cpu-jit]               |      0.151774 |        0.060143 |     -60.4% |
| test_eval[vgg16-cpu-jit]                         |      0.415365 |        0.203798 |     -50.9% |
```

**Comment by gmagogsfm** [comment]
@penguinwu Seems related to IR manipulatiblity work you brought up.

**Comment by penguinwu** [comment]
> @penguinwu Seems related to IR manipulatiblity work you brought up.

Right, I was just reading about that.

**Comment by eellison** [comment]
@pinzhenx sorry for the delay, things move a little slowly around holiday and end-of-half. We met and discussed this PR and we think we need to do a little more research and homework on our end before moving forward.

Out of curiosity - what was it about the profiled executor that didn't work with the fuser ? I know what thing that could limit was need for graph cleanup - I added https://github.com/pytorch/pytorch/pull/50221 but obviously it would be great to do this in a more comprehensive manner.

**Comment by pinzhenx** [comment]
@eellison Thank you for following up

> what was it about the profiled executor that didn't work with the fuser?

We just enabled the profiled executor and verified some of the tests.

I found that Conv-Bn fusion is no longer a problem (due to logic of the profile node insertion has been changed by [this PR](https://github.com/pytorch/pytorch/commit/97a2918a0708d06fe7610912dc698cb08b88a8f4?branch=97a2918a0708d06fe7610912dc698cb08b88a8f4&diff=split#diff-b1bacad4061ec4366b57fc98129776bb74927a3a5de64a1284bde637d4cb7a69R198)). But we got another problem in Conv-BN-Relu. Since relu is supported by TE, a profile node will be inserted before relu, which prevents the fusion of conv-bn-relu.

It's only one of the problems we've found so far. Later, we will further investigate the compatibility of onednn fuser and profiling_executor along with TE fuser.

**Comment by fzhao3** [comment]
This is being proposed as a Beta feature for PT2.0, and here is the full list of PRs for this feature.

- #76622
- #85591
- #82491

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
